src/shared/telemetry/spans.ts

```diff
@@ -0,0 +1,254 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import globals from '../extensionGlobals'
+
+import type { AsyncLocalStorage as AsyncLocalStorageClass } from 'async_hooks'
+import {
+    definitions,
+    Metadata,
+    Metric,
+    MetricBase,
+    MetricDefinition,
+    MetricName,
+    MetricShapes,
+    TelemetryBase,
+} from './telemetry'
+import { getTelemetryReason, getTelemetryResult } from '../errors'
+
+const AsyncLocalStorage: typeof AsyncLocalStorageClass =
+    require('async_hooks').AsyncLocalStorage ??
+    class<T> {
+        readonly #store: T[] = []
+        #disabled = false
+
+        public disable() {
+            this.#disabled = true
+        }
+
+        public getStore() {
+            return this.#disabled ? undefined : this.#store[0]
+        }
+
+        public run<R>(store: T, callback: (...args: any[]) => R, ...args: any[]): R {
+            this.#disabled = false
+            this.#store.unshift(store)
+
+            try {
+                const result = callback(...args)
+                if (result instanceof Promise) {
+                    return result.finally(() => this.#store.shift()) as unknown as R
+                }
+                this.#store.shift()
+                return result
+            } catch (err) {
+                this.#store.shift()
+                throw err
+            }
+        }
+
+        public exit<R>(callback: (...args: any[]) => R, ...args: any[]): R {
+            const saved = this.#store.shift()
+
+            try {
+                const result = callback(...args)
+                if (result instanceof Promise) {
+                    return result.finally(() => saved !== undefined && this.#store.unshift(saved)) as unknown as R
+                }
+                saved !== undefined && this.#store.unshift(saved)
+                return result
+            } catch (err) {
+                saved !== undefined && this.#store.unshift(saved)
+                throw err
+            }
+        }
+
+        public enterWith(store: T): void {
+            // XXX: you need hooks into async resource lifecycles to implement this correctly
+            this.#store.shift()
+            this.#store.unshift(store)
+        }
+    }
+
+function getValidatedState(state: Partial<MetricBase>, definition: MetricDefinition) {
+    const missingFields: string[] = []
+
+    for (const [k, v] of Object.entries(state)) {
+        if (definition.requiredMetadata.includes(k) && v === undefined) {
+            missingFields.push(k)
+        }
+    }
+
+    if (missingFields.length !== 0) {
+        return Object.assign({ missingFields }, state)
+    }
+
+    return { ...state }
+}
+
+export class TelemetrySpan {
+    #startTime: Date | undefined = undefined
+    private readonly state: Partial<MetricBase> = {}
+    private readonly definition = definitions[this.name] ?? {
+        unit: 'None',
+        passive: true,
+        requiredMetadata: [],
```

Comment:
nit: could just name this `required`.

side-note: "metadata" has seemed like a misleading name because it's not really meta--it's actually the data.

Reply:
Originally the data was 'value' but that's hardly ever used. So yeah it should be called just 'data'. Less confusing too.